### PR TITLE
Update THC Prepare to use QROAMClean

### PR DESCRIPTION
**Pull request by fdmalone**
Use QROAMClean for THC prepare. Hit some issues with resource estimation notebook and fixes a bug with LCUBlockEncoding junk registers which may be sided (at least I think this is the desired behaviour).

BlackBoxPrepare does not work with sided registers either so I removed this from the resource estimation notebook (xref #1381)

I updated the call graph slightly to remove CirqGateAsBloq references and makes sure the decomposition matches the call graph.

I removed the PrepareTHC circuit diagram as the new QROAMClean bloq results in way too many junk registers making it unsightly (or really it just fails by default: #1380)

### CODE REVIEW

**Comment by fdmalone** [comment]
xref #1381 and #1380

**Comment by fdmalone** [comment]
summoning @tanujkhattar

File: qualtran/bloqs/chemistry/thc/prepare.py [hunk 1]
```diff
@@ -361,124 +368,138 @@ def selection_registers(self) -> Tuple[Register, ...]:
...
-        alt_mu, alt_nu = alt_mn
-        s, theta, alt_theta, alt_mu, alt_nu, keep = bb.add(
-            qroam,
-            selection=s,
-            target0_=theta,
-            target1_=alt_theta,
-            target2_=alt_mu,
-            target3_=alt_nu,
-            target4_=keep,
+        # TODO remove these
```
Comment (tanujkhattar) [inline qualtran/bloqs/chemistry/thc/prepare.py:456]: Can you add some more context for this TODO ?

File: qualtran/bloqs/chemistry/thc/prepare.py [hunk 2]
```diff
@@ -361,124 +368,138 @@ def selection_registers(self) -> Tuple[Register, ...]:
...
         cost_4 = (OnEach(self.keep_bitsize, Hadamard()), 1)
         cost_5 = (LessThanEqual(self.keep_bitsize, self.keep_bitsize), 2)
         cost_6 = (CSwap(nmu), 3)
-        cost_7 = (Toffoli(), 1)
-        return {cost_1, cost_2, cost_3, cost_4, cost_5, cost_6, cost_7}
+        cost_7 = (MultiControlX(cvs=(0, 1)), 1)
+        cost_8 = (XGate(), 2)
+        cost_9 = (CZ(), 2)
+        cost_10 = (Hadamard(), 3)
+        return {cost_1, cost_2, cost_3, cost_4, cost_5, cost_6, cost_7, cost_8, cost_9, cost_10}
```
Comment (tanujkhattar) [inline qualtran/bloqs/chemistry/thc/prepare.py:502]: nit: Please use a dictionary instead of set, since set representation will be deprecated soon.

**Review by tanujkhattar — APPROVED** [review]
LGTM % nits

File: qualtran/bloqs/chemistry/resource_estimation.ipynb [hunk 3]
```diff
@@ -158,35 +140,22 @@
...
-    "        if isinstance(k, CSwap):\n",
-    "            toffolis += v * k.bitsize\n",
-    "        elif isinstance(k, TGate):\n",
-    "            toffolis += v // 4\n",
-    "    return toffolis\n",
+    "def get_toffoli_counts(bloq: Bloq) -> SymbolicInt:\n",
+    "    return get_cost_value(bloq, QECGatesCost(), generalizer=generalize_cswap_approx).total_t_and_ccz_count(ts_per_rotation=0)['n_ccz']\n",
     "\n",
     "num_toff = get_toffoli_counts(block_encoding_bloq)\n",
+    "print(num_toff)\n",
```
[the remark this answers is not in the document]
Comment (fdmalone) [reply qualtran/bloqs/chemistry/resource_estimation.ipynb:158]: Removed.

File: qualtran/bloqs/chemistry/thc/prepare.py [hunk 4]
```diff
@@ -361,124 +368,138 @@ def selection_registers(self) -> Tuple[Register, ...]:
...
+            Register('alt_nu', QAny(bitsize=self.num_mu.bit_length()), side=Side.RIGHT),
+            Register('keep', QAny(bitsize=self.keep_bitsize), side=Side.RIGHT),
+        )
+
+    @cached_property
+    def qroam_extra_target_registers(self) -> Tuple[Register, ...]:
+        """Extra registers required for QROAMClean."""
+        return tuple(
+            Register(
+                name=f'jnk_{reg.name}',
```
[the remark this answers is not in the document]
Comment (fdmalone) [reply qualtran/bloqs/chemistry/thc/prepare.py:394]: No, replaced.

File: qualtran/bloqs/chemistry/thc/prepare.py (hunk 1 above)
Comment (fdmalone) [reply qualtran/bloqs/chemistry/thc/prepare.py:456]: Oh, I think this is done. Removed.

File: qualtran/bloqs/chemistry/thc/prepare.py (hunk 2 above)
Comment (fdmalone) [reply qualtran/bloqs/chemistry/thc/prepare.py:502]: Done.

**Pull request merged**